File: lib/unicode-to-brf.ts
// Mapping: Unicode Braille glyph → ASCII Braille glyph
const UNICODE_TO_BRF: Record<string, string> = {
  '⠀': ' ',   // Braille blank (U+2800)
  ' ': ' ',    // Normal space
  '\u00A0': ' ',  // Non-breaking space (NBSP)
  '⠮': '!', '⠐': '"', '⠼': '#', '⠫': '$', '⠩': '%',
  '⠯': '&', '⠄': "'", '⠷': '(', '⠾': ')', '⠡': '*', '⠬': '+',
  '⠠': ',', '⠤': '-', '⠨': '.', '⠌': '/', '⠴': '0', '⠂': '1',
  '⠆': '2', '⠒': '3', '⠲': '4', '⠢': '5', '⠖': '6', '⠶': '7',
  '⠦': '8', '⠔': '9', '⠱': ':', '⠰': ';', '⠣': '<', '⠿': '=',
  '⠜': '>', '⠹': '?', '⠈': '@', '⠁': 'A', '⠃': 'B', '⠉': 'C',
  '⠙': 'D', '⠑': 'E', '⠋': 'F', '⠛': 'G', '⠓': 'H', '⠊': 'I',
  '⠚': 'J', '⠅': 'K', '⠇': 'L', '⠍': 'M', '⠝': 'N', '⠕': 'O',
  '⠏': 'P', '⠟': 'Q', '⠗': 'R', '⠎': 'S', '⠞': 'T', '⠥': 'U',
  '⠧': 'V', '⠺': 'W', '⠭': 'X', '⠽': 'Y', '⠵': 'Z', '⠪': '[',
  '⠳': '\\', '⠻': ']', '⠘': '^', '⠸': '_'
};

/**
 * Convert Unicode Braille text to BRF ASCII encoding (TypeScript fallback implementation).
 * Handles space normalization (normal space, NBSP, Braille blank) and removes unintended consecutive symbols.
 */
function unicodeToBrfFallback(unicodeText: string): string {
  // Normalize all types of spaces
  let text = unicodeText.replace(/\u00A0/g, ' ').replace(/⠀/g, ' ');
  
  // Remove unintended consecutive Braille symbols like ⠹⠹
  text = text.replace(/(⠹)\1+/g, '$1');
  
  // Convert each character
  const brfChars: string[] = [];
  for (const ch of text) {
    if (ch in UNICODE_TO_BRF) {
      brfChars.push(UNICODE_TO_BRF[ch]);
    } else {
      // Check if it's a Braille character (U+2800 to U+28FF)
      const codePoint = ch.codePointAt(0);
      if (codePoint !== undefined && codePoint >= 0x2800 && codePoint <= 0x28FF) {
        // It's a Braille character but not in our mapping - try to map it
        // For unmapped Braille characters, we'll use '?' as fallback
        console.warn(`⚠ Warning: Unmapped Braille symbol '${ch}' (U+${codePoint.toString(16).toUpperCase().padStart(4, '0')})`);
        brfChars.push('?');
      } else {
        // Non-Braille character - keep as is (for Tamil text headers, etc.)
        brfChars.push(ch);
      }
    }
  }
  
  return brfChars.join('');
}

/**
 * Convert Unicode Braille text to BRF ASCII encoding using Python API.
 * Falls back to TypeScript implementation if API is unavailable.
 */
export async function unicodeToBrf(unicodeText: string): Promise<string> {
  // Try to use Python API first
  try {
    const response = await fetch('/api/brf', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ unicodeText }),
    });

    if (response.ok) {
      const data = await response.json();
      return data.brfText;
    } else {
      // API failed (500 or 503) - use TypeScript fallback
      // 503 indicates service unavailable (e.g., Python not available on Vercel)
      const status = response.status;
      if (status === 503 || status === 500) {
        // Silently use fallback - this is expected on Vercel
        return unicodeToBrfFallback(unicodeText);
      }
      // For other errors, log but still use fallback
      console.warn(`Python BRF conversion API failed with status ${status}, using TypeScript fallback`);
      return unicodeToBrfFallback(unicodeText);
    }
  } catch (error) {
    // Network error or API not available (e.g., on Vercel without Python), use fallback
    // Don't log errors in production as this is expected behavior
    if (process.env.NODE_ENV === 'development') {
      console.warn('Python BRF conversion API unavailable, using TypeScript fallback:', error);
    }
    return unicodeToBrfFallback(unicodeText);
  }
}

/**
 * Download text content as a file
 */
export function downloadTextFile(content: string, filename: string, mimeType: string = 'text/plain') {
  const blob = new Blob([content], { type: mimeType });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

File: lib/unicode-to-brf.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Unicode Braille to BRF ASCII conversion utility.
Can be used as a standalone script or imported as a module.
"""

import sys
import re

# Mapping: Unicode Braille glyph → ASCII Braille glyph
UNICODE_TO_BRF = {
    '\u2800': ' ',   # Braille blank (U+2800)
    ' ': ' ',        # Normal space
    '\u00A0': ' ',   # Non-breaking space (NBSP)
    '\u282E': '!',   # ⠮
    '\u2810': '"',   # ⠐
    '\u283C': '#',   # ⠼
    '\u282B': '$',   # ⠫
    '\u2829': '%',   # ⠩
    '\u282F': '&',   # ⠯
    '\u2804': "'",   # ⠄
    '\u2837': '(',   # ⠷
    '\u283E': ')',   # ⠾
    '\u2821': '*',   # ⠡
    '\u282C': '+',   # ⠬
    '\u2820': ',',   # ⠠
    '\u2824': '-',   # ⠤
    '\u2828': '.',   # ⠨
    '\u280C': '/',   # ⠌
    '\u2834': '0',   # ⠴
    '\u2802': '1',   # ⠂
    '\u2806': '2',   # ⠆
    '\u2812': '3',   # ⠒
    '\u2832': '4',   # ⠲
    '\u2822': '5',   # ⠢
    '\u2816': '6',   # ⠖
    '\u2836': '7',   # ⠶
    '\u2826': '8',   # ⠦
    '\u2814': '9',   # ⠔
    '\u2831': ':',   # ⠱
    '\u2830': ';',   # ⠰
    '\u2823': '<',   # ⠣
    '\u283F': '=',   # ⠿
    '\u281C': '>',   # ⠜
    '\u2839': '?',   # ⠹
    '\u2808': '@',   # ⠈
    '\u2801': 'A',   # ⠁
    '\u2803': 'B',   # ⠃
    '\u2809': 'C',   # ⠉
    '\u2819': 'D',   # ⠙
    '\u2811': 'E',   # ⠑
    '\u280B': 'F',   # ⠋
    '\u281B': 'G',   # ⠛
    '\u2813': 'H',   # ⠓
    '\u280A': 'I',   # ⠊
    '\u281A': 'J',   # ⠚
    '\u2805': 'K',   # ⠅
    '\u2807': 'L',   # ⠇
    '\u280D': 'M',   # ⠍
    '\u281D': 'N',   # ⠝
    '\u2815': 'O',   # ⠕
    '\u280F': 'P',   # ⠏
    '\u281F': 'Q',   # ⠟
    '\u2817': 'R',   # ⠗
    '\u280E': 'S',   # ⠎
    '\u281E': 'T',   # ⠞
    '\u2825': 'U',   # ⠥
    '\u2827': 'V',   # ⠧
    '\u283A': 'W',   # ⠺
    '\u282D': 'X',   # ⠭
    '\u283D': 'Y',   # ⠽
    '\u2835': 'Z',   # ⠵
    '\u282A': '[',   # ⠪
    '\u2833': '\\',  # ⠳
    '\u283B': ']',   # ⠻
    '\u2818': '^',   # ⠘
    '\u2838': '_',   # ⠸
}


def unicode_to_brf(unicode_text: str) -> str:
    """
    Convert Unicode Braille text to BRF ASCII encoding.
    
    Args:
        unicode_text: Unicode Braille text string
        
    Returns:
        BRF ASCII encoded string
    """
    # Normalize all types of spaces
    text = unicode_text.replace('\u00A0', ' ').replace('\u2800', ' ')
    
    # Remove unintended consecutive Braille symbols like ⠹⠹
    text = re.sub(r'(\u2839)\1+', r'\1', text)
    
    # Convert each character
    brf_chars = []
    for ch in text:
        if ch in UNICODE_TO_BRF:
            brf_chars.append(UNICODE_TO_BRF[ch])
        else:
            # Check if it's a Braille character (U+2800 to U+28FF)
            code_point = ord(ch)
            if 0x2800 <= code_point <= 0x28FF:
                # It's a Braille character but not in our mapping
                print(f"⚠ Warning: Unmapped Braille symbol '{ch}' (U+{code_point:04X})", file=sys.stderr)
                brf_chars.append('?')
            else:
                # Non-Braille character - keep as is (for Tamil text headers, etc.)
                brf_chars.append(ch)
    
    return ''.join(brf_chars)


def main():
    """Main function for command-line usage."""
    if len(sys.argv) < 2:
        # Read from stdin if no arguments
        unicode_text = sys.stdin.read()
    else:
        # Use first argument as input
        unicode_text = sys.argv[1]
    
    brf_content = unicode_to_brf(unicode_text)
    print(brf_content, end='')


if __name__ == '__main__':
    main()
